File: CodeWars/StringCalculator/InfixToReversePolishParser.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace CodeWars.StringCalculator
{
    public class InfixToReversePolishParser
    {
        #region Fields

        private string _convertedTask = "";
        private string _taskLineBefore = "";
        private readonly Stack<string> _operationsBuffer = new Stack<string>();
        private bool _isCurrentNegative = false;

        #endregion


        #region Parser

        public string Parse(string task)
        {
            string[] taskInLines = Regex.Split(PrepareTask(task).Replace(" ", ""), "(?<=[-+*/()^])|(?=[-+*/()^])").Where(i => i != "").ToArray();

            foreach (var taskLine in taskInLines)
            {
                if (IsNegative(taskLine))
                {
                    _isCurrentNegative = true;
                    continue;
                }

                if (!CalculatorUtilities.IsOperatorOrBracket(taskLine))
                {
                    _convertedTask += (IsNegativeNumber(taskLine) ? "-" : "") + (taskLine + " ");
                    _isCurrentNegative = false;
                }

                else
                    OperationsPriorityHandler(taskLine);

                _taskLineBefore = taskLine;
            }

            if (_operationsBuffer.Count > 0) OperationsPriorityHandler(isLast: true);

            return _convertedTask;
        }

        #endregion


        #region Helpers

        private static string HandleNegativeBrackets(string task)
        {
            IEnumerable<string> taskInBrackets = Regex.Split(task.Replace(" ", ""), "([-]{1}[(]{1}|[(]{1}|[)]{1})").Where(i => i != "").ToArray();

            string taskWithWrappedNegativeBrackets = "";
            var bracketsCache = new Stack<string>();
            string lastLine = "";

            // Handling negative brackets
            foreach (string currentLine in taskInBrackets)
            {
                if (CalculatorUtilities.IsOperator(lastLine.LastOrDefault().ToString()) && currentLine.Equals("-("))
                {
                    bracketsCache.Push("-(");
                    taskWithWrappedNegativeBrackets += "(0";
                }

                if (lastLine.Equals("-(") && currentLine.Equals("-("))
                    taskWithWrappedNegativeBrackets += "0";

                else if (currentLine.Equals("(") && bracketsCache.Count > 0)
                    bracketsCache.Push("(");

                else if (currentLine.Equals(")") && bracketsCache.Count > 0 && bracketsCache.Pop().Equals("-("))
                    taskWithWrappedNegativeBrackets += ")";


                lastLine = currentLine;
                taskWithWrappedNegativeBrackets += currentLine;
            }

            return taskWithWrappedNegativeBrackets;
        }

        private static string HandleSingleNumbersInBracketsAndRepeatingBrackets(string taskWithWrappedNegativeBrackets)
        {
            IEnumerable<string> taskInSingleNumbersInsideBrackets =
                Regex.Split(taskWithWrappedNegativeBrackets.Replace(" ", ""), "([(]{1}|[)]{1})").Where(i => i != "");

            string taskWithHandledSingleNumbersInsideBrackets = "";

            string lastLine = "";

            // Handling single numbers inside brackets
            foreach (string currentLine in taskInSingleNumbersInsideBrackets)
            {
                if (lastLine.Equals("(") && currentLine.Equals("("))
                    taskWithHandledSingleNumbersInsideBrackets += "0+";

                double number;
                if (double.TryParse(currentLine, out number))
                {
                    if (number < 0)
                        taskWithHandledSingleNumbersInsideBrackets += "0";
                    else
                        taskWithHandledSingleNumbersInsideBrackets += "0+";
                }

                lastLine = currentLine;
                taskWithHandledSingleNumbersInsideBrackets += currentLine;
            }

            return taskWithHandledSingleNumbersInsideBrackets;
        }

        private static string PrepareTask(string task)
        {
            return HandleSingleNumbersInBracketsAndRepeatingBrackets(HandleNegativeBrackets(task));
        }

        private void AddLastItemToConvertedTaskString(string item)
        {
            _convertedTask += item == "(" || item == ")" ? "" : item + " ";
        }

        private void OperationsPriorityHandler(string currentLine = "", bool emptyTheBracket = false, bool isLast = false)
        {
            if (isLast)
            {
                AddLastItemToConvertedTaskString(_operationsBuffer.Pop());
                OperationsPriorityHandler(currentLine, isLast: _operationsBuffer.Count > 0);
            }

            else if (currentLine.Equals(")") || emptyTheBracket)
            {
                if (_operationsBuffer.Count < 1) return;

                string currentItem = emptyTheBracket ? currentLine : _operationsBuffer.Pop();

                AddLastItemToConvertedTaskString(currentItem);

                if (_operationsBuffer.Count < 1) return;

                string nextItem = _operationsBuffer.Pop();

                if (!nextItem.Equals("(")) OperationsPriorityHandler(nextItem, emptyTheBracket: true);
            }

            else if (_operationsBuffer.Count == 0) _operationsBuffer.Push(currentLine);

            else if (CalculatorUtilities.IsCurrentOperationPriorityBigger(currentLine, _operationsBuffer.Peek())) _operationsBuffer.Push(currentLine);

            else
            {
                AddLastItemToConvertedTaskString(_operationsBuffer.Pop());
                OperationsPriorityHandler(currentLine);
            }
        }

        private bool IsNegative(string currentLine)
        {
            return (_taskLineBefore == "" || _taskLineBefore == "(" || CalculatorUtilities.IsOperator(_taskLineBefore)) && currentLine.Equals("-");
        }

        private bool IsNegativeNumber(string currentLine)
        {
            return _isCurrentNegative && !CalculatorUtilities.IsOperatorOrBracket(currentLine);
        }

        #endregion
    }
}

File: CodeWarsJS/2kyuEvaluateMathematicalExpression/InfixToReversePolishParser.js
"use strict";

module.exports = class InfixToReversePolishParser {
    constructor() {
        this.calculatorUtilities = require("./CalculatorUtilities.js");

        this.convertedTask = "";

        this.taskLineBefore = "";
        this.operationsBuffer = [];

        this.isCurrentNegative = false;
    }

    parse(task) {
        let taskInLines = this.prepareTask(task).replace(/ /g, "").split(/([-+*/()^])/).filter(i => i !== "");

        taskInLines.forEach(taskLine => {
            if (this.checkIsNegative(taskLine)) {
                this.isCurrentNegative = true;
                return;
            }

            if (!this.calculatorUtilities.isOperatorOrBracket(taskLine)) {
                this.convertedTask += (this.checkIsNegativeNumber(taskLine) ? "-" : "") + (taskLine + " ");
                this.isCurrentNegative = false;
            }

            else
                this.operationsPriorityHandler(taskLine, false, false);

            this.taskLineBefore = taskLine;
        });

        if (this.operationsBuffer.length > 0) this.operationsPriorityHandler("", false, true);

        return this.convertedTask;
    }


    // Utilities

    handleNegativeBrackets(task) {
        let taskInBrackets = task.replace(/ /g, "").split(/([-]{1}[(]{1}|[(]{1}|[)]{1})/).filter(i => i !== "");

        let taskWithWrappedNegativeBrackets = "";
        let bracketsCache = [];
        let lastLine = "";

        // handling negative brackets
        taskInBrackets.forEach(currentLine => {

            if (lastLine !== "" && this.calculatorUtilities.isOperator(lastLine[lastLine.length - 1]) && currentLine === "-(") {
                bracketsCache.push("-(");
                taskWithWrappedNegativeBrackets += "(0";
            }

            if (lastLine === "-(" && currentLine === "-(")
                taskWithWrappedNegativeBrackets += "0";

            else if (currentLine === "(" && bracketsCache.length > 0)
                bracketsCache.push("(");

            else if (currentLine === ")" && bracketsCache.length > 0 && bracketsCache.pop() === "-(")
                taskWithWrappedNegativeBrackets += ")";

            lastLine = currentLine;
            taskWithWrappedNegativeBrackets += currentLine;
        });

        return taskWithWrappedNegativeBrackets;
    }

    handleSingleNumbersInBracketsAndRepeatingBrackets(taskWithWrappedNegativeBrackets) {
        let taskInSingleNumbersInsideBrackets = taskWithWrappedNegativeBrackets.replace(/ /g, "").split(/([(]{1}|[)]{1})/).filter(i => i !== "");
        let taskWithHandledSingleNumbersInsideBrackets = "";

        let lastLine = "";

        // Handling single numbers inside brackets
        taskInSingleNumbersInsideBrackets.forEach(currentLine => {
            if (lastLine === "(" && currentLine === "(")
                taskWithHandledSingleNumbersInsideBrackets += "0+";

            if (this.isFloat(currentLine)) {
                if (parseFloat(currentLine) < 0)
                    taskWithHandledSingleNumbersInsideBrackets += "0";
                else
                    taskWithHandledSingleNumbersInsideBrackets += "0+";
            }

            lastLine = currentLine;
            taskWithHandledSingleNumbersInsideBrackets += currentLine;
        });

        return taskWithHandledSingleNumbersInsideBrackets;
    }

    isFloat(number) {
        return (number[0] === "-" || !isNaN(number[0])) && number.length > 1 && !number.substr(1).match(/[-+*/()^]/);
    }

    prepareTask(task) {
        return this.handleSingleNumbersInBracketsAndRepeatingBrackets(this.handleNegativeBrackets(task));
    }

    addLastItemToConvertedTaskString(item) {
        this.convertedTask += item === "(" || item === ")" ? "" : item + " ";
    }

    checkIsNegative(currentLine) {
        return (this.taskLineBefore === "" || this.taskLineBefore === "(" || this.calculatorUtilities.isOperator(this.taskLineBefore))
            && currentLine === "-";
    }

    checkIsNegativeNumber(currentLine) {
        return this.isCurrentNegative && !this.calculatorUtilities.isOperatorOrBracket(currentLine);
    }

    operationsPriorityHandler(currentLine, emptyTheBracket, isLast) {
        if (isLast) {
            this.addLastItemToConvertedTaskString(this.operationsBuffer.pop());
            this.operationsPriorityHandler(currentLine, false, this.operationsBuffer.length > 0);
        }

        else if (currentLine === ")" || emptyTheBracket) {
            if (this.operationsBuffer.length < 1) return;

            let currentItem = emptyTheBracket ? currentLine : this.operationsBuffer.pop();

            this.addLastItemToConvertedTaskString(currentItem);

            if (this.operationsBuffer.length < 1) return;

            let nextItem = this.operationsBuffer.pop();

            if (nextItem !== "(") this.operationsPriorityHandler(nextItem, true, false);
        }

        else if (this.operationsBuffer.length === 0) this.operationsBuffer.push(currentLine);

        else if (this.calculatorUtilities.isCurrentOperationPriorityBigger(currentLine, this.operationsBuffer[this.operationsBuffer.length - 1])) this.operationsBuffer.push(currentLine);

        else {
            this.addLastItemToConvertedTaskString(this.operationsBuffer.pop());
            this.operationsPriorityHandler(currentLine, false, false);
        }
    }

}

